packages/firestore/test/integration/api/database.test.ts

```diff
@@ -541,59 +541,29 @@ apiDescribe('Database', (persistence: boolean) => {
     });
   });
 
-  apiDescribe('onSnapshotsInSync listener', () => {
-    function createDocumentListener(
-      target: firestore.DocumentReference,
-      callback: (snapshot: firestore.DocumentSnapshot) => void
-    ): void {
-      target.onSnapshot(snap => {
-        callback(snap);
-      });
-    }
+  it('onSnapshotsInSync fires after listeners are in sync', () => {
+    const testDocs = {
+      a: { foo: 1 }
+    };
+    return withTestCollection(persistence, testDocs, async coll => {
+      let events: string[] = [];
+      let done = new Deferred<void>();
+      const doc = coll.doc('a');
 
-    function createQueryListener(
-      target: firestore.Query,
-      callback: (snapshot: firestore.QuerySnapshot) => void
-    ): void {
-      target.onSnapshot(snap => {
-        callback(snap);
+      doc.onSnapshot(snap => {
+        events.push('doc');
       });
-    }
-
-    it('fires after local listeners fire', () => {
-      const testDocs = {
-        a: { foo: 1 }
-      };
-      return withTestCollection(persistence, testDocs, async coll => {
-        const events: string[] = [];
-        const done = new Deferred<void>();
-        const doc = coll.doc('a');
-
-        createDocumentListener(doc, snapshot => {
-          events.push('doc1');
-        });
-        createQueryListener(coll.where('foo', '>', 0), snapshot => {
-          events.push('query');
-        });
-        createDocumentListener(doc, snapshot => {
-          events.push('doc2');
-        });
-        onSnapshotsInSync(doc.firestore, () => {
-          events.push('snapshots-in-sync');
-          if (events.length === 5) {
-            expect(events).to.deep.equal([
-              'snapshots-in-sync',
-              'doc1',
-              'doc2',
-              'query',
-              'snapshots-in-sync'
-            ]);
-            done.resolve();
-          }
-        });
-
-        await done.promise;
+      onSnapshotsInSync(doc.firestore, () => {
+        events.push('snapshots-in-sync');
+        done.resolve();
       });
+      await done.promise;
+      done = new Deferred<void>();
+      events = [];
+
+      await doc.set({ foo: 3 });
+      await done.promise;
+      expect(events).to.deep.equal(['doc', 'snapshots-in-sync']);
```

Comment:
While this test is probably deterministic in JavaScript since it is single-threaded, it will be racey on Android / iOS since there is a race between onSnapshot() getting an initial event from the server and it getting the event from your local set() call.  If the server event comes first, you could end up with ['doc', 'snapshots-in-sync', 'doc', 'snapshots-in-sync']

There are a variety of ways to make this non-racey.  One way would be to explicitly wait for the initial snapshot event.  It's not super elegant, but something like:

```
  it.only('onSnapshotsInSync fires after listeners are in sync', () => {
    const testDocs = {
      a: { foo: 1 }
    };
    return withTestCollection(persistence, testDocs, async coll => {
      let events: string[] = [];
      const doc = coll.doc('a');
      const gotInitialSnapshot = new Deferred<void>();
      doc.onSnapshot(snap => {
        events.push('doc');
        gotInitialSnapshot.resolve();
      });
      await (gotInitialSnapshot.promise);
      events = [];

      const done = new Deferred<void>();
      onSnapshotsInSync(doc.firestore, () => {
        events.push('snapshots-in-sync');
        if (events.length === 3) {
          // We should have an initial snapshots-in-sync event, then snapshot event for set(), then
          // another event to indicate we're in sync again.
          expect(events).to.deep.equal(['snapshots-in-sync', 'doc', 'snapshots-in-sync']);
          done.resolve();
        }
      });

      await doc.set({ foo: 3 });
      await done.promise;
    });
  });
```